apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/(analysis)/responses/actions.ts

```diff
@@ -14,17 +20,45 @@ export const resolveResponseNoteAction = async (responseNoteId: string) => {
 };
 
 export const deleteResponseAction = async (responseId: string) => {
+  const session = await getServerSession(authOptions);
+  if (!session) throw new AuthorizationError("Not authorized");
+
+  const isAuthorized = await canUserAccessResponse(session.user.id, responseId);
+  if (!isAuthorized) throw new AuthorizationError("Not authorized");
+
   return await deleteResponse(responseId);
 };
 
 export const createTagAction = async (environmentId: string, tagName: string) => {
+  const session = await getServerSession(authOptions);
+  if (!session) throw new AuthorizationError("Not authorized");
+
+  const isAuthorized = await hasUserEnvironmentAccess(session.user.id, environmentId);
+  if (!isAuthorized) throw new AuthorizationError("Not authorized");
+
   return await createTag(environmentId, tagName);
 };
 
 export const addTagToResponeAction = async (responseId: string, tagId: string) => {
+  console.log("addTagToResponeAction", responseId, tagId);
+
+  const session = await getServerSession(authOptions);
+  if (!session) throw new AuthorizationError("Not authorized");
+
+  const isAuthorizedForTag = await canUserAccessTag(session.user.id, tagId);
+  const isAuthorizedForResponse = await canUserAccessResponse(session.user.id, responseId);
+  if (!isAuthorizedForTag || !isAuthorizedForResponse) throw new AuthorizationError("Not authorized");
+
   return await addTagToRespone(responseId, tagId);
 };
 
 export const removeTagFromResponseAction = async (responseId: string, tagId: string) => {
+  const session = await getServerSession(authOptions);
+  if (!session) throw new AuthorizationError("Not authorized");
+
+  const isAuthorizedForTag = await canUserAccessTag(session.user.id, tagId);
+  const isAuthorizedForResponse = await canUserAccessResponse(session.user.id, responseId);
```

Comment:
Since `tagOnResponse` is an own model and service (strange name I know, we need to change this in the future):
Shouldn't be there an auth `canUserAccessTagOnResponse`?